kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/AdminManager.java

```diff
@@ -439,10 +445,30 @@ private void handleUpdatePartitionsAsync(String topic,
     public void setBrokers(Map<String, Set<Node>> newBrokers) {
         brokersCacheLock.writeLock().lock();
         try {
+            setControllerId(newBrokers);
```

Comment:
BTW, based on my example code before, this method can move out of the locked block.

```java
    public void setBrokers(Map<String, Set<Node>> newBrokers) {
        setControllerId(newBrokers);
        brokersCacheLock.writeLock().lock();
        try {
            this.brokersCache = newBrokers;
        } finally {
            brokersCacheLock.writeLock().unlock();
        }
    }

    private void setControllerId(Map<String, Set<Node>> newBrokers) {
        final Map<String, Integer> listenerToControllerId = /* create a local map using a local variable `newBrokers` */;
        synchronized (this.listenerToControllerId) { // local the field to make it thread safe, there's no need to protect it by `brokersCacheLock`.
            this.listenerToControllerId.clear();
            this.listenerToControllerId.putAll(listenerToControllerId);
        }
    }

    private final Map<String, Integer> listenerToControllerId = Maps.newHashMap();
```

Reply:
you can see this issue: https://issues.apache.org/jira/browse/KAFKA-7019